blockchain/ethereum/ethereum.go

```diff
@@ -0,0 +1,202 @@
+	Network                 Network `json:"network"` // One of "ganache", "ropsten", or "mainnet" ("mainnet" is not current supported)
+	URI                     string  `json:"uri"`
+	RepublicTokenAddress    string  `json:"republicTokenAddress"`
+	DarknodeRegistryAddress string  `json:"darknodeRegistryAddress"`
+	RenLedgerAddress        string  `json:"renLedgerAddress"`
+	RenExAccountsAddress    string  `json:"renExAccountsAddress"`
+}
+
+// Conn contains the client and the contracts deployed to it
+type Conn struct {
+	RawClient *ethrpc.Client
+	Client    *ethclient.Client
+	Config    Config
+}
+
+// Connect to a URI.
+func Connect(config Config) (Conn, error) {
+	if config.URI == "" {
+		switch config.Network {
+		case NetworkGanache:
+			config.URI = "http://localhost:8545"
+		case NetworkKovan:
+			config.URI = "https://kovan.infura.io"
+		case NetworkRopsten:
+			config.URI = "https://ropsten.infura.io"
+		default:
+			return Conn{}, fmt.Errorf("cannot connect to %s: unsupported", config.Network)
+		}
+	}
+	if config.RepublicTokenAddress == "" {
+		switch config.Network {
+		case NetworkGanache:
+			config.RepublicTokenAddress = RepublicTokenAddressOnGanache.String()
+		case NetworkKovan:
+			config.RepublicTokenAddress = RepublicTokenAddressOnKovan.String()
+		case NetworkRopsten:
+			config.RepublicTokenAddress = RepublicTokenAddressOnRopsten.String()
+		default:
+			return Conn{}, fmt.Errorf("cannot connect to %s: unsupported", config.Network)
+		}
+	}
+	if config.DarknodeRegistryAddress == "" {
+		switch config.Network {
+		case NetworkGanache:
+			config.DarknodeRegistryAddress = DarknodeRegistryAddressOnGanache.String()
+		case NetworkKovan:
+			config.DarknodeRegistryAddress = DarknodeRegistryAddressOnKovan.String()
+		case NetworkRopsten:
+			config.DarknodeRegistryAddress = DarknodeRegistryAddressOnRopsten.String()
+		default:
+			return Conn{}, fmt.Errorf("cannot connect to %s: unsupported", config.Network)
+		}
+	}
+	if config.RenLedgerAddress == "" {
+		switch config.Network {
+		case NetworkGanache:
+			config.RenLedgerAddress = RenLedgerAddressOnGanache.String()
+		case NetworkKovan:
+			config.RenLedgerAddress = RenLedgerAddressOnKovan.String()
+		case NetworkRopsten:
+			config.RenLedgerAddress = RenLedgerAddressOnRopsten.String()
+		default:
+			return Conn{}, fmt.Errorf("cannot connect to %s: unsupported", config.Network)
+		}
+	}
+
+	if config.RenExAccountsAddress == "" {
+		switch config.Network {
+		case NetworkGanache:
+			config.RenExAccountsAddress = RenExAccountsAddressOnGanache.String()
+		case NetworkKovan:
+			config.RenExAccountsAddress = RenExAccountsAddressOnKovan.String()
+		case NetworkRopsten:
+			panic("not deployed yet")
+		default:
+			return Conn{}, fmt.Errorf("cannot connect to %s: unsupported", config.Network)
+		}
+	}
+	// Fixme: hardcode the renex contract for now,
+	// need to update the address in therir config files
+	config.RenExAccountsAddress = RenExAccountsAddressOnKovan.String()
```

Comment:
Do we want to have all the contract addresses in the config file or coed in the darknode ?

Reply:
It was agreed that having them directly programmed into the Darknodes was better for rolling out updates. Going through and changing that now.